src/html/Quoting.cc

```diff
@@ -9,81 +9,74 @@
 #include "squid.h"
 #include "html/Quoting.h"
 
-/*
- *  HTML defines these characters as special entities that should be quoted.
- */
-static struct {
-    unsigned char code;
-    const char *quote;
-} htmlstandardentities[] =
+#include <array>
+#include <cstring>
+
+class HtmlSpecialCharacters {
+private:
+    std::array<const char *, 256> escapeMap = {};
 
+public:
+    HtmlSpecialCharacters();
+    ~HtmlSpecialCharacters();
+    const char *operator[](unsigned char ch) const { return escapeMap[ch]; }
+};
+HtmlSpecialCharacters::HtmlSpecialCharacters()
 {
-    /* NOTE: The quoted form MUST not be larger than 6 character.
-     * see close to the MemPool commend below
+    static std::array<std::pair<unsigned char, const char *>, 5> const escapePairs = {
+        std::make_pair('<', "&lt;"),
+        std::make_pair('>', "&gt;"),
+        std::make_pair('"', "&quot;"),
+        std::make_pair('&', "&amp;"),
+        std::make_pair('\'', "&apos;")
+    };
+    const size_t maxEscapeLength = 7;
+    /* Encode control chars just to be on the safe side, and make
+     * sure all 8-bit characters are encoded to protect from buggy
+     * clients
      */
-    {
-        '<', "&lt;"
-    },
-    {
-        '>', "&gt;"
-    },
-    {
-        '"', "&quot;"
-    },
-    {
-        '&', "&amp;"
-    },
-    {
-        '\'', "&#39;"
-    },
-    {
-        0, NULL
+    for (uint32_t ch = 1; ch < 256; ++ch) {
+        if ((ch <= 0x1F || ch >= 0x7f) && ch != '\n' && ch != '\r' && ch != '\t') {
+            escapeMap[ch] = static_cast<char *>(xcalloc(maxEscapeLength, 1));
+            snprintf(const_cast<char*>(escapeMap[ch]), sizeof escapeMap[ch], "&#%d;", static_cast<int>(ch));
+        }
     }
-};
+    for (auto &pair: escapePairs) {
+        escapeMap[pair.first] = static_cast<char *>(xcalloc(maxEscapeLength, 1));
+        xstrncpy(const_cast<char*>(escapeMap[pair.first]), pair.second, maxEscapeLength);
+    }
+}
+HtmlSpecialCharacters::~HtmlSpecialCharacters()
+{
+    for (auto &escape: escapeMap) {
+        xfree(escape);
+    }
+}
 
-/*
- *  html_do_quote - Returns a static buffer containing the quoted
- *  string.
- */
 char *
 html_quote(const char *string)
 {
-    static char *buf = nullptr;
+    // for syntax, see https://html.spec.whatwg.org/#character-references
+    static const auto htmlSpecialCharacters = HtmlSpecialCharacters();
+
+    static char *buf;
     static size_t bufsize = 0;
     const char *src;
     char *dst;
-    int i;
 
-    /* XXX This really should be implemented using a MemPool, but
-     * MemPools are not yet available in lib...
+    /* XXX This really should be implemented using a MemPool
      */
-    if (!buf || strlen(string) * 6 > bufsize) {
+    if (buf == nullptr || strlen(string) * 6 > bufsize) {
         xfree(buf);
         bufsize = strlen(string) * 6 + 1;
-        buf = static_cast<char *>(xcalloc(bufsize, 1));
+        buf = static_cast<char*>(xcalloc(bufsize, 1));
```

Comment:
Please undo these unnecessary, partially controversial (and arguably unwanted) changes, starting from the XXX comment and ending with `         buf = static_cast<char...` below it.

Reply:
done